File: exercises/ddd_infrastructure_layer/java/src/main/java/application/services/ApproveInitialApplicationService.java
package application.services;

import java.util.UUID;

import com.google.inject.Inject;

import domain.application.ApplicationId;
import domain.application.ApplicationRepository;

public class ApproveInitialApplicationService {
    private ApplicationRepository repo;

    @Inject
    public ApproveInitialApplicationService(ApplicationRepository repo) {
        this.repo = repo;
    }

    public void execute(ApproveInitialApplicationCommand cmd) throws Exception {
        var applicationId = new ApplicationId(cmd.applicationId());
        var application = repo.getById(applicationId);
        if (application == null) {
            throw new RuntimeException("Application not found for id " + applicationId.getValue());
        }

        try {
            application.approveInitialApplication();
        } catch (Exception e) {
            throw new RuntimeException("Failed to approve initial application: " + e.getMessage());
        }

        repo.save(application);
    }

    public static record ApproveInitialApplicationCommand(UUID applicationId) {}
}

File: exercises/ddd_infrastructure_layer/typescript/src/application/services/ApproveInitialApplicationService.ts
import { ApplicationId } from "../../domain/application/ApplicationId";
import type { ApplicationRepository } from "../../domain/application/ApplicationRepository";

export interface ApproveInitialApplicationCommand {
  applicationId: string; // UUIDv4
}

export class ApproveInitialApplicationService {
  private repo: ApplicationRepository;

  constructor(repo: ApplicationRepository) {
    this.repo = repo;
  }

  async execute(cmd: ApproveInitialApplicationCommand) {
    const id = new ApplicationId(cmd.applicationId);
    const application = await this.repo.getById(id);

    if (!application) {
      throw new Error(`Application not found for id ${cmd.applicationId}`);
    }

    try {
      application.approveInitialApplication();
    } catch (e) {
      throw new Error(`Failed to approve initial application: ${(e as Error).message}`);
    }

    await this.repo.save(application);
  }
}
